### ISSUE TITLE
inference 하는 법

### ISSUE THREAD

**Issue by Trailblazer-Yoo** [BoocampMonster/korean-relation-extraction#23]
@yhkee0404 @hundredeuk2 @chanmuzi

```python
def main(config):
    seed_setting(config.train.seed)
    # 데이터셋 로드 클래스를 불러옵니다.
    
    assert torch.cuda.is_available(), "GPU를 사용할 수 없습니다."
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    
    tokenizer = AutoTokenizer.from_pretrained(config.model.model_name)

    data = pd.read_csv(config.data.test_path)
    with open('data/dict_num_to_label.pkl', 'rb') as f:
        dict_num_to_label = pickle.load(f)
        
    test = getattr(DataModule, config.model.data_class)(
        mode = "test",
        data=data,
        tokenizer=tokenizer,
        entity_marker_mode= config.data.get('entity_marker_mode'),
        max_length=config.train.max_length)  
    test_dataloader = DataLoader(test, batch_size=1, pin_memory=True, shuffle=False)
    
    # 모델 아키텍처를 불러옵니다.
    model = getattr(Model, config.model.model_class)(
        model_name = config.model.model_name,
        num_labels=30,
        dropout_rate = config.model.dropout_rate,
        add_token_num = config.data.get('entity_marker_num')
        ).to(device)
    checkpoint = torch.load(f'./save/{config.model.saved_dir}/epoch:4_model.pt') ### 🔥🔥epoch 수정 필요🔥🔥 ###
    model.load_state_dict(checkpoint)
    
    model = model.to(device)
    model.eval()
    all_preds = []
    all_probs = []
    with torch.no_grad():
        for batch in tqdm(test_dataloader):
            if 'mask' in config.model.saved_dir: ## 🔥🔥 save의 해당 디렉토리에 submission.csv파일이 저장됨 🔥🔥 ##
                output = model(batch["input_ids"].to(device),
                           batch["attention_mask"].to(device),
                           entity_mask1 = batch['entity_mask1'],
                           entity_mask2 = batch['entitymaskd2'])
            elif config.data.get('entity_marker_mode'):
                output = model(batch["input_ids"].to(device),
                           batch["attention_mask"].to(device),
                           entity_embed1 = batch['entity_embed1'],
                           entity_embed2 = batch['entity_embed2'])
            else:
                output = model(batch["input_ids"].to(device),
                        batch["attention_mask"].to(device))
            output = ein.rearrange(output, '1 class -> class')

            probs = F.softmax(output, dim=-1).detach().cpu().numpy()
            preds = np.argmax(probs, axis=-1).item()
            
            all_preds.append(preds)
            all_probs.append(probs.tolist())
    
    ## make csv file with predicted answer
    #########################################################
    # 아래 directory와 columns의 형태는 지켜주시기 바랍니다.
    output_df = pd.DataFrame(data['id'])
    output_df['pred_label'] = all_preds
    output_df['probs'] = all_probs
    output_df['pred_label'] = output_df['pred_label'].apply(lambda x: dict_num_to_label[x])

    output_df.to_csv(f'save/{config.model.saved_dir}/submission.csv', index=False) # 최종적으로 완성된 예측한 라벨 csv 파일 형태로 저장.
    print('추론 완료')

if __name__=='__main__':
    torch.cuda.empty_cache()
    parser = argparse.ArgumentParser()
    parser.add_argument('--config', type=str, default='')
    args, _ = parser.parse_known_args()
    
    file = 'koelectra_entity_marker_punct_tokens' ## 🔥🔥 돌리고 싶은 config 파일 이름을 넣어주세요(for loop도 가능) 🔥🔥 ##
    config_w = OmegaConf.load(f'./configs/{file}.yaml')
    print(f'사용할 수 있는 GPU는 {torch.cuda.device_count()}개 입니다.')
    
    main(config_w)
```

### 위의 코드를 그대로 복붙해서 🔥🔥 이 있는 부분을 참고
1. wandb 로그 결과 가장 결과가 좋은 pt 파일을 수정
2. file 이름을 수정

**Comment by Trailblazer-Yoo** [comment]
그렇게 써도 됨
`file = 'koelectra_entity_marker_punct_tokens' ## 🔥🔥 돌리고 싶은 config 파일 이름을 넣어주세요(for loop도 가능)`
여기를 지우고
그 밑줄에
{file} 부분을 args.config로 바꿔주기만 하면 됨
추론은 config 넣는것도 귀찮아서 그냥 저렇게 쓴것 뿐임~

**Comment by hundredeuk2** [comment]
`test = getattr(DataModule, config.model.data_class)(
        mode = "test",
        data=data,
        tokenizer=tokenizer,
        entity_marker_mode= config.data.get('entity_marker_mode'),
        mask_mode=config.data.mask_mode,
        max_length=config.train.max_length)  
    test_dataloader = DataLoader(test, batch_size=16, pin_memory=True, shuffle=False)`
여기서 `mask_mode`랑 `max_length` yaml파일에 없어서 에러뜨는데 삭제해도 됩니까?

**Comment by Trailblazer-Yoo** [comment]
@hundredeuk2 max_length는 지우면 안될거 같고 mask_mode는 지워주셈
mask_mode는 내가 모델이나 데이터셋을 python 파일로 분할하지 않았을때 파라미터라
저걸 지우는게 맞는듯 max_length는 들어가있을건데
